Sources/GRPCCore/Transport/InProcessClientTransport.swift

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2023, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+/// An in-process implementation of a ``ClientTransport``.
+public struct InProcessClientTransport: ClientTransport {
+  private enum State: Sendable {
+    case unconnected(InProcessServerTransport)
+    case connected(InProcessServerTransport)
+    case closed
+  }
+
+  public typealias Inbound = RPCAsyncSequence<RPCResponsePart>
+  public typealias Outbound = RPCWriter<RPCRequestPart>.Closable
+
+  public var retryThrottle: RetryThrottle
+
+  private let executionConfigurations: ClientRPCExecutionConfigurationCollection
+  private let state: LockedValueBox<State>
+
+  public init(
+    server: InProcessServerTransport,
+    executionConfigurations: ClientRPCExecutionConfigurationCollection
+  ) {
+    self.retryThrottle = .init(maximumTokens: 10, tokenRatio: 0.1)
+    self.executionConfigurations = executionConfigurations
+    self.state = .init(.unconnected(server))
+  }
+
+  /// Establish and maintain a connection to the remote destination.
+  ///
+  /// Maintains a long-lived connection, or set of connections, to a remote destination.
+  /// Connections may be added or removed over time as required by the implementation and the
+  /// demand for streams by the client.
+  ///
+  /// Implementations of this function will typically create a long-lived task group which
+  /// maintains connections. The function exits when all open streams have been closed and new connections
+  /// are no longer required by the caller who signals this by calling ``close()``, or by cancelling the
+  /// task this function runs in.
+  ///
+  /// - Parameter lazily: This parameter is ignored in this implementation.
+  public func connect(lazily: Bool) async throws {
```

Comment:
This returns quickly but should only return when the connection is closed and all streams are handled.

You'll need to manage a bit more state here and I suspect the way to go is to do the connect and create and store a continuation which indicates when the `connect` can return.